### PR TITLE
Selki/ff 3514 bandit parameters polling only when needed

### PR DESCRIPTION
**Pull request by maya-the-mango**
---
labels: mergeable
---
[//]: #  (Link to the issue corresponding to this chunk of work)
Fixes: #__issue__

## Motivation and Context
[//]: #  Requesting bandits only when needed (when modelVersions differ) saves bandwidth! :)

## Description
[//]: # Before requesting new bandits I do a simple comparison of loaded model version vs the newest model version contained in UFC

## How has this been tested?
[//]: # 

[//]: # (OPTIONAL)
[//]: # (Add one or multiple labels: enhancement, refactoring, bugfix)

### CODE REVIEW

**Comment by maya-the-mango** [comment]
2 unit tests are failing due to changes in test json file that happened before I even forked the repo. I've fixed them already though! But in this [PR](https://github.com/Eppo-exp/js-client-sdk-common/pull/128) so we can ignore those here. :))

File: src/configuration-requestor.ts
Comment (typotter) [inline src/configuration-requestor.ts:50]: Can you check `banditReferences` at line 30/36
`const flagsHaveBandits = Object.keys(configResponse.banditReferences ?? {}).length > 0;`

`bandits` is deprecated and replaced with `banditReferences`. If we're able to 🪓 the use of `bandits` in this PR, that'd be great, but it might be too much noise.

File: src/configuration-requestor.ts
Comment (typotter) [inline src/configuration-requestor.ts:53]: what do you think about moving this check before `requiresBanditModelConfigurationStoreUpdate`? Then you wouldn't have to make this check again in that method. It would also save having to do a fetch just to throw the error because there's nowhere to store the fetched data.

You could also pass the current loaded bandits as a parameter. I think the methods will be more testable as well if you can replace `this.` references with params.

File: src/configuration-requestor.ts
Comment (typotter) [inline src/configuration-requestor.ts:67]: 💪

File: src/configuration-requestor.ts
Comment (typotter) [inline src/configuration-requestor.ts:88]: Could this be set when bandits are loaded "set"? I think it's worth it to frontload & store this in memory instead of on each request (typical polling interval is 30s)

File: src/configuration-requestor.ts
Comment (typotter) [inline src/configuration-requestor.ts:92]: 💪

**Review by typotter — CHANGES_REQUESTED** [review]
looking great and definitely on the right track!
Just some comments I hope may make your testing easier.
Thanks for the lift in this SDK!

File: src/http-client.ts
Comment (typotter) [inline src/http-client.ts:29]: Can this field be removed now?

File: src/configuration-requestor.ts
Comment (typotter) [inline src/configuration-requestor.ts:85]: consider skipping the setter and just using
`this.banditModelVersions = this.getLoadedBanditModels(...)`

File: src/http-client.ts
Comment (maya-the-mango) [reply src/http-client.ts:29]: Yes! I will do just that

File: src/configuration-requestor.spec.ts
Comment (aarsilv) [inline src/configuration-requestor.spec.ts:211]: Ideally, our test checks that it hasn't changed the scenario as well. You can do this by updating the mock between (this would be my instinct) or changing the response based on call count (like we do in our "retries initial configuration request" tests).

Psudocode:
```
init SDK
expect bandits to have been fetched
fetch (UFC bandit references are not updated)
expect no additional bandit fetches
fetch (bandit references model versions shrunk to a _subset_)
expect no additional bandit fetches
poll for updated ufc (bandit references model versions a _superset_)
expect bandits to have been fetched
```

File: src/configuration-requestor.ts
Comment (aarsilv) [inline src/configuration-requestor.ts:95]: 👍

**Review by aarsilv — COMMENTED** [review]
This is looking good! Only request is a more beefed up test for this!

File: src/configuration-requestor.spec.ts
Comment (maya-the-mango) [inline src/configuration-requestor.spec.ts:142]: @aarsilv @typotter I'm thinking about moving the polling tests to a isolated test suite and then reset `fetchSpy` to default state  using `afterAll()`

```javascript
describe('ConfigurationRequestor', {
     describe('Bandits with flags', {
        describe('Bandits polling', {
            afterAll(() => {
                initiateFetchSpy(defaultResponseMockGenerator);
            })
            // polling tests here
          });
     });
});
```

File: src/configuration-requestor.spec.ts
Comment (maya-the-mango) [reply src/configuration-requestor.spec.ts:246]: cc: @typotter

**Review by maya-the-mango — COMMENTED** [review]
I beefed up the tests! But I've never done this kind of fetch mocking shenanigans before in js, so I'm not 100% confident in the approach I took: passing mock response generator function to fetchSpy initializer and the updateUFC flag which serves as a semi global variable - I hate using this but nothing better comes to my mind tbh.

File: src/configuration-requestor.spec.ts
Comment (aarsilv) [reply src/configuration-requestor.spec.ts:246]: dirty but seems to do the trick 🤷‍♂️

File: src/configuration-requestor.spec.ts
Comment (aarsilv) [inline src/configuration-requestor.spec.ts:279]: e.g., you should do another fetch before and make sure bandits is not fetched--ideally with a _subset_

File: src/configuration-requestor.spec.ts
Comment (aarsilv) [reply src/configuration-requestor.spec.ts:246]: consider an else where you delete an existing bandit. That way you can check once with bandits A,B, then another with bandits A, then lastly with bandits A,C. The first and third should fetch bandits (but not the middle)--that is what we want to test!

File: src/configuration-requestor.spec.ts
Comment (typotter) [reply src/configuration-requestor.spec.ts:142]: always a fan of better organization. maybe in a stacked or follow-on PR would be better

**Review by typotter — COMMENTED** [review]
Just about there!
Can't agree with you more about the awkwardness and repetition encountered when testing bandits. 
Ides for how to refactor are most welcome!

File: package.json
Comment (aarsilv) [inline package.json:3]: Since this is not adding new (from the users perspective) functionality but instead just adjusting something existing (similar to a bugfix) let's just bump the patch number by one: `4.3.1`. (if this is out of date, rebase on latest main)

File: src/configuration-requestor.spec.ts
Comment (aarsilv) [inline src/configuration-requestor.spec.ts:233]: these will be hoised to the top-level describe. Consider putting them in either in it's new nested `describe()` block along with the three tests that use it.

File: src/configuration-requestor.spec.ts
Comment (aarsilv) [inline src/configuration-requestor.spec.ts:371]: Love the comment!

**Review by aarsilv — APPROVED** [review]
Thanks for iterating!

Approving, but before merging you'll want to do three things:
1. Rebase on latest main and ensure `package.json` is just a patch (last number) bump
2. Install this locally and put together a corresponding change and PR for the node SDK to make sure it works upstream as intended
3. Get the sign off from @typotter who requested changes

**Review by typotter — APPROVED** [review]
Great stuff. Going to see some solid improvement on our fastly request volumes with this.
LGTM, pending the version change and other comments from @aarsilv

File: src/configuration-requestor.spec.ts
Comment (maya-the-mango) [reply src/configuration-requestor.spec.ts:371]: 🙌

**Pull request merged**
